lib/xapi-stdext-date/date.ml

```diff
@@ -44,50 +45,54 @@ type iso8601 = Ptime.t
   match x |> Ptime.of_rfc3339 |> Ptime.rfc3339_error_to_msg with
   | Error (`Msg e) -> invalid_arg (Printf.sprintf "date.ml:of_string: %s" e)
   | Ok (t, tz, _)  -> match tz with
-                      | None | Some 0 -> t
+                      | None | Some 0 -> (t, UTC)
                       | Some _        -> invalid_arg (Printf.sprintf "date.ml:of_string: %s" x)
 
-let to_string t =
-  Ptime.to_rfc3339 ~tz_offset_s:0 (* to ensure Z printed, rather than +00:00 *) t |>
-  Astring.String.filter (fun char -> char <> '-') (* remove dashes for backwards compatibility *)
+let to_string (t, zone) =
+  let t_string =
+    (* to ensure Z printed, rather than +00:00 *)
+    Ptime.to_rfc3339 ~tz_offset_s:0 t
+  in
+  match zone with
+  (* in all cases, remove '-' for backwards compatibility. for localtime we must also remove 'Z' *)
+  | UTC   -> Astring.String.filter (function '-' -> false | _ -> true) t_string
+  | Local -> Astring.String.filter (function '-' | 'Z' -> false | _ -> true) t_string
 
-let of_float x =
-  let time = Unix.gmtime x in
-  Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02dZ"
-    (time.Unix.tm_year+1900)
-    (time.Unix.tm_mon+1)
-    time.Unix.tm_mday
-    time.Unix.tm_hour
-    time.Unix.tm_min
-    time.Unix.tm_sec |> of_string
-
-(* Convert tm in localtime to calendar time, x *)
-let to_float_localtime x =
-  let datetime_to_float y mon d h min s =
-    fst Unix.(mktime { tm_year = y - 1900;
-                       tm_mon = mon - 1;
-                       tm_mday = d;
-                       tm_hour = h;
-                       tm_min = min;
-                       tm_sec = s;
-                       (* These are ignored: *)
-                       tm_wday = 0; tm_yday = 0; tm_isdst = true;
-                     })
+let of_float s =
+  let t = match Ptime.of_float_s s with
+          | None -> invalid_arg (Printf.sprintf "date.ml:of_float: %f" s)
+          | Some t -> t
   in
-  let ((y, mon, d), ((h, min, s), _)) = Ptime.to_date_time x in
-  datetime_to_float y mon d h min s
+  (t, UTC)
+
+let _offset_by t offset =
+  let offset_int = match offset with None -> 0 | Some i -> i in
+  let span = Ptime.Span.of_int_s offset_int in
+  match Ptime.add_span t span with
+  | None ->
+      invalid_arg
+        (Printf.sprintf
+           "date.ml:localtime: failed to adjust time (%s) with offset (%i)"
+           (Ptime.to_rfc3339 t) offset_int)
+  | Some now_local ->
+      now_local
+
+let localtime () =
+  (_offset_by (Ptime_clock.now ()) (Ptime_clock.current_tz_offset_s ()), Local)
```

Comment:
I think it'd be simpler if we do this at to_string time. This way the Ptime.t doesn't always store UTC, which is confusing.

Reply:
My understanding is that `Ptime.t` doesn't have to store UTC, it's just on the UTC timeline (otherwise why would we be provided the function `Ptime.add_span`?)

>The type for picosecond precision POSIX timestamps in the range [Ptime.min;Ptime.max]. Note that POSIX timestamps, >and hence values of this type, are by definition always on the UTC timeline.

(To me it seems more complicated to carry around the offset)

Reply:
You can of course add a value to a UTC timestamp and get another UTC timestamp (e.g. `n hours before now, or n hours after now`). If we actually merge the offset into the Ptime.t then it won't be in UTC anymore, which is confusing if someone comes and reads the code later on: the Ptime docs say it is UTC, but we use it in a different way.

Reply:
>  If we actually merge the offset into the Ptime.t then it won't be in UTC anymore

It's still 'UTC' (i.e. on the UTC timeline like the docs say):

```
# let now = Ptime_clock.now ();;
# let Some offset = Ptime_clock.current_tz_offset_s ();;
# let offset_span = Ptime.Span.of_int_s offset;;
# let Some adjusted = Ptime.add_span now offset_span;;
# adjusted |> Ptime.to_rfc3339;;
- : string = "2020-08-10T10:17:35-00:00"
```
comparing with `now`:

```
# now |> Ptime.to_rfc3339;;
- : string = "2020-08-10T09:17:35-00:00"
```

>Another possibility is to store this as Ptime.date * Ptime.time insteafd of Ptime.t which then properly includes the offset, but without us doing anything complicated to store it.

I would not be a proponent of this because then we'd have to handle offsets in `to_float` and `of_float` (I suppose we could keep the variant)

Reply:
In the example you've shown `"2020-08-10T10:17:35-00:00"` is UTC, but it doesn't represent the current time in UTC. If you try to use it as UTC it'll be off by the amount of offset you added (it is in "local" timezone now).

Reply:
In its current form you don't support Local time in `to_float`, you could still do the same in the `date * time` version: if the offset is nonzero you raise the invalid exception.

Reply:
I see what you mean now in that it would more clear to store it as +HH:MM. I'll make that change and use `date * time` so we can avoid the weird string manipulation